text/0000-style-fn-decl.md

```diff
@@ -0,0 +1,274 @@
+(which is underspecified in any case).
+
+For reference the current text in the style guide states:
+
+> For multiline function signatures, each new line should align with the
+> first parameter. Multiple parameters per line are permitted:
+> 
+> ``` rust
+> fn frobnicate(a: Bar, b: Bar,
+>               c: Bar, d: Bar)
+>               -> Bar {
+>     ...
+> }
+> 
+> fn foo<T: This,
+>        U: That>(
+>        a: Bar,
+>        b: Bar)
+>        -> Baz {
+>     ...
+> }
+> ```
+
+I believe that this does not work very well, since when scanning it is hard to
+distinguish arguments from type parameters and the return type. It also does not
+specify anything for where clauses.
+
+# Detailed design
+
+## Examples
+
+```
+// a comment
+#[an_attribute]
+#[another_attribute]
+pub unsafe extern "C" fn foo<A, B, C>(x: SomeType,
+                                      y: Vec<B>,
+                                      z: SomeOtherType<B, C>)
+    -> ReturnType<A>
+    where A: Foo + Bar,
+          B: Baz<B, C>,
+{
+```
+
+```
+// a comment
+fn foo<A, B, C>(x: SomeType,
+                y: Vec<B>,
+                z: SomeOtherType<B, C>)
+    -> ReturnType<A>
+{
+```
+
+```
+// a comment
+fn foo(x: SomeType,
+       y: Vec<B>,
+       z: SomeOtherType<B, C>)
+{
+```
+
+```
+// a comment
+pub unsafe extern "C" fn foo<A, B, C>(x: SomeType,
+                                      y: Vec<B>,
+                                      z: SomeOtherType<B, C>)
+    where A: Foo + Bar,
+          B: Baz<B, C>,
+{
+```
+
+```
+fn foo<T>(x: T, y: A) -> T where T: Bar {
+
+// alternative
+fn foo<T>(x: T, y: A) -> T
+    where T: Bar
+{
+```
+
+```
+fn foo<T>(x: T, y: A) -> T {
+```
+
+## Rules
+
+* Comments first, then attributes, one line each, then the rest of the signature on a new line;
+* if the whole signature fits on one line (excluding comments and attributes), do it;
+* if the signature requires more than one line,
+  - each argument gets its own line, argument names are aligned;
+  - the return type gets its own line, with block indent + 4;
+  - the where clause gets its own line, aligned with the return type (or block indent + 4, if no return type);
+  - each predicate in a where clause gets its own line, with the start of the predicates aligned;
+  - the opening brace gets a new line at block indent.
+
+Prefer not to split type parameters. If it is necessary, then the opening
+parenthesis and opening angle bracket should be aligned. Type parameter names
+should be aligned (but prefer one line if possible):
+
+```
+// Prefer this
+fn foo<A, B, C>
+      (x: SomeType,
+       y: Vec<B>,
+       z: SomeOtherType<B, C>)
+{
+
+// If we need more room (but prefer a where clause, if it is bounds causing the problem):
+fn foo<A,
+       B,
+       C>
+      (x: SomeType,
+       y: Vec<B>,
+       z: SomeOtherType<B, C>)
+{
+```
+
+If more space is necessary for arguments, then the opening parenthesis may be on
```

Comment:
I like how in struct constructors the arguments all look the same (`name: Type,`), no matter if they are the first or the last. I suggest doing the same for function generics and arguments. Imo this increases readability significantly and allows easy adding/removing of function generics and arguments (add/remove a line). Therefor I suggest that the opening parenthesis is always on the same line, and if the previous styles take up more than ~40 chars of width, use the style below